src/ray/raylet/node_manager.cc

```diff
@@ -596,26 +639,31 @@ void NodeManager::ProcessNodeManagerMessage(TcpClientConnection &node_manager_cl
   node_manager_client.ProcessMessages();
 }
 
-void NodeManager::ScheduleTasks() {
-  auto policy_decision = scheduling_policy_.Schedule(
-      cluster_resource_map_, gcs_client_->client_table().GetLocalClientId(),
-      remote_clients_);
+// Make a placement decision for placeable tasks given the resource_map provided.
+// Perform task state transition and task forwarding based on the placement decision.
+void NodeManager::ScheduleTasks(
+    std::unordered_map<ClientID, SchedulingResources> &resource_map) {
```

Comment:
Questions:

1. Are you adding `resource_map` as an argument as a way of giving a "hint" to this method (so that it can only consider a single remote machine when you get a heartbeat from that machine)?

2. `SchedulingResources` in its current form doesn't seem to contain as much information as we'd like to have in order to make the scheduling decision, e.g., it doesn't let us distinguish static capacity from dynamic capacity from queue size, right? Not that we should do all of that in this PR, but how would we do that down the road? E.g., by augmenting `SchedulingResources` or by passing more things into this method?

Reply:
1. good question. Here I am leveraging the generality of our SchedulingPolicy interface, by calling it to make placement decisions with the cluster resource map restricted to the heartbeating node alone. This, in essence, should reuse the same policy to implement targetted spillover to the heartbeating node.
2. If I understand correctly, when you say "dynamic capacity" and "static capacity" you are referring to enumerated (e.g., GPUs) and simple, capacity-based (e.g., CPUs) resource types, correct? SchedulingResources is a container class that operates on resource sets, each of which represents available, total, and load information (added in this PR). The ability to differentiate between enumerated and simple resource types should've been implemented by introducing a polymorphic Resource type with EnumeratedResource and SimpleResource deriving from it. Then ResourceSet serves as the container of such polymorphic `Resource` objects. I plan to attempt this refactoring as soon as I finish this PR.

Reply:
By "dynamic" and "static", I meant "available" and "total". I didn't realize that `SchedulingResources` contains all three pieces of information (available/total/queued). I assumed it only had two out of the three.

Reply:
Also, I think we currently only have "enumerated" resources, so there may be no need to have both an `EnumeratedResource` type and a `SimpleResource` type.

Reply:
a. ah yes, we are definitely tracking the equivalent of "static" and "dynamic" resources as well as the load information, each represented as a ResourceSet inside SchedulingResources.
b. I certainly anticipate that we will have a mixture of resource types that consist of resources we will wish to enumerate and assign IDs on dispatch and resources we only need to keep track of overall capacity for (e.g., memory).